### ISSUE TITLE
Application gets blocked on db connection acquisition if more than 10 concurrent clients connect

### ISSUE THREAD

**Issue by irumiha** [pauljamescleary/scala-pet-store#149]
I'm not sure if it's about this application or this is a property of all http4s / doobie / hikari applications. It seems that something gets stuck when there are more clients connected than available connections in the connection pool.

How to reproduce: 

1. do sbt dist (to remove the sbt environment as a factor)
2. unpack and run the app 
3. run `siege http://localhost:8080/users`

Use top, htop or whatever to monitor system activity. You will notice nothing is happening. After 30 seconds exceptions start popping up:

```
[scala-execution-context-global-33] ERROR o.h.s.service-errors - Error servicing request: GET /users from 127.0.0.1 
java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:676)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:190)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:155)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)
	at doobie.util.transactor$Transactor$fromDataSource$FromDataSourceUnapplied.$anonfun$apply$5(transactor.scala:260)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:87)
	at cats.effect.internals.IORunLoop$.startCancelable(IORunLoop.scala:41)
	at cats.effect.internals.IOBracket$BracketStart.run(IOBracket.scala:86)
	at cats.effect.internals.Trampoline.cats$effect$internals$Trampoline$$immediateLoop(Trampoline.scala:70)
	at cats.effect.internals.Trampoline.startLoop(Trampoline.scala:36)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.super$startLoop(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.$anonfun$startLoop$1(TrampolineEC.scala:93)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:85)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.startLoop(TrampolineEC.scala:93)
	at cats.effect.internals.Trampoline.execute(Trampoline.scala:43)
	at cats.effect.internals.TrampolineEC.execute(TrampolineEC.scala:44)
	at cats.effect.internals.IOBracket$BracketStart.apply(IOBracket.scala:72)
	at cats.effect.internals.IOBracket$BracketStart.apply(IOBracket.scala:52)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

Stop the `siege` tool and observe the report:

```
** SIEGE 4.0.4
** Preparing 25 concurrent users for battle.
The server is now under siege...^C
Lifting the server siege...
Transactions:		           0 hits
Availability:		        0.00 %
Elapsed time:		       71.92 secs
Data transferred:	        0.00 MB
Response time:		        0.00 secs
Transaction rate:	        0.00 trans/sec
Throughput:		        0.00 MB/sec
Concurrency:		        8.22
Successful transactions:           0
Failed transactions:	          34
Longest transaction:	       61.85
Shortest transaction:	        0.00
```

Now if you restart the app and run siege again, but limit the number of concurrent "users" to 10: 

```
$ siege -c10 http://localhost:8080/users
** SIEGE 4.0.4
** Preparing 10 concurrent users for battle.
The server is now under siege...^C
Lifting the server siege...
Transactions:		      134101 hits
Availability:		      100.00 %
Elapsed time:		       11.06 secs
Data transferred:	        0.26 MB
Response time:		        0.00 secs
Transaction rate:	    12124.86 trans/sec
Throughput:		        0.02 MB/sec
Concurrency:		        9.26
Successful transactions:      134101
Failed transactions:	           0
Longest transaction:	        0.03
Shortest transaction:	        0.00
```

Everything works OK. Why 10? This application does not configure the Hikari connection pool so defaults are used.

I am entirely unable to debug this but I thought I'd share my findings.

**Comment by zakpatterson** [comment]
Hey thanks so much for reporting. I reviewed your PR, happy to discuss there or in gitter.

**Comment by pauljamescleary** [comment]
@irumiha thanks for finding this.

10 is the max number of connections in the connection pool (unless specified otherwise) - https://github.com/brettwooldridge/HikariCP

Since we are using the same thread pool throughout, eventually if you have more requesters than connections, those additional concurrent requests will block/wait for a connection to get freed up.  Too many users and then too many will block and eventually time out waiting for a connection.

We have double trouble as we use the same execution context for servicing requests AND making database calls. 

Realistically, changing the max pool size feels like it should help out.  Definitely using a separate service execution context from your database calls which are blocking on IO will help as well (that you already found out).

**Comment by zakpatterson** [comment]
Thanks @irumiha for your fix in #150
